mavros/src/plugins/sys_time.cpp

```diff
@@ -289,30 +354,122 @@ class SystemTimePlugin : public plugin::PluginBase {
 		UAS_FCU(m_uas)->send_message_ignore_drop(tsync);
 	}
 
-	void add_timesync_observation(int64_t offset_ns, uint64_t observation_time)
+	void add_timesync_observation(int64_t offset_ns, uint64_t local_time_ns, uint64_t remote_time_ns)
 	{
-		int64_t dt = time_offset_ns - offset_ns;
+		uint64_t now_ns = ros::Time::now().toNSec();
 
-		if (std::abs(dt) > 10000000) {		// 10 millisecond skew
-			time_offset_ns = offset_ns;	// hard-set it.
-			m_uas->set_time_offset(time_offset_ns);
+		// Calculate the round trip time (RTT) it took the timesync packet to bounce back to us from remote system
+		uint64_t rtt_ns = now_ns - local_time_ns;
+
+		// Calculate the difference of this sample from the current estimate
+		uint64_t deviation = llabs(int64_t(time_offset) - offset_ns);
+
+		if (rtt_ns < max_rtt_sample * 1000000ULL) {	// Only use samples with low RTT
+			if (sync_converged() && (deviation > max_deviation_sample * 1000000ULL)) {
+				// Increment the counter if we have a good estimate and are getting samples far from the estimate
+				high_deviation_count++;
+
+				// We reset the filter if we received consecutive samples which violate our present estimate.
+				// This is most likely due to a time jump on the offboard system.
+				if (high_deviation_count > max_cons_high_deviation) {
+					ROS_ERROR_NAMED("time", "TM : Time jump detected. Resetting time synchroniser.");
+
+					// Reset the filter
+					reset_filter();
+
+					// Reset diagnostics
+					dt_diag.clear();
+					dt_diag.set_timestamp(remote_time_ns);
+				}
+			} else {
+				// Filter gain scheduling
+				if (!sync_converged()) {
+					// Interpolate with a sigmoid function
+					float progress = float(sequence) / convergence_window;
+					float p = 1.0f - expf(0.5f * (1.0f - 1.0f / (1.0f - progress)));
+					filter_alpha = p * filter_alpha_final + (1.0f - p) * filter_alpha_initial;
+					filter_beta = p * filter_beta_final + (1.0f - p) * filter_beta_initial;
+				} else {
+					filter_alpha = filter_alpha_final;
+					filter_beta = filter_beta_final;
+				}
+
+				// Perform filter update
+				add_sample(offset_ns);
+
+				// Save time offset for other components to use
+				m_uas->set_time_offset(sync_converged() ? time_offset : 0);
+
+				// Increment sequence counter after filter update
+				sequence++;
+
+				// Reset high deviation count after filter update
+				high_deviation_count = 0;
+
+				// Reset high RTT count after filter update
+				high_rtt_count = 0;
+			}
+		} else {
+			// Increment counter if round trip time is too high for accurate timesync
+			high_rtt_count++;
+
+			if (high_rtt_count > max_cons_high_rtt) {
+				ROS_WARN_NAMED("time", "TM : RTT too high for timesync: %0.2f ms", rtt_ns / 1000000.0);
+				// Reset counter to rate-limit warnings
+				high_rtt_count = 0;
+			}
```

Comment:
ROS already have THROTTLEd logger. setting not needed.

Reply:
Should've made comment clearer. The intention is to not just rate-limit, but rather to inform the user when the RTT is high _n times in a row_. This would point to an issue with the system which needs to be fixed by the user.